flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/DefaultCheckpointPlanCalculator.java

```diff
@@ -147,11 +147,7 @@ private void checkAllTasksInitiated() throws CheckpointException {
      */
     private void checkTasksStarted(List<Execution> toTrigger) throws CheckpointException {
         for (Execution execution : toTrigger) {
-            if (execution.getState() == ExecutionState.CREATED
-                    || execution.getState() == ExecutionState.RECOVERING
-                    || execution.getState() == ExecutionState.SCHEDULED
-                    || execution.getState() == ExecutionState.DEPLOYING) {
-
+            if (execution.getState() != ExecutionState.RUNNING) {
```

Comment:
Didn't @gaoyunhaii say that we also need to trigger tasks if they are in state `FINISHED`? I think here he said it: https://issues.apache.org/jira/browse/FLINK-22003?focusedCommentId=17312877&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17312877

Reply:
Hi Till, very sorry for I ignored some details for the first comment, in realistic although we might taken checkpoint when the tasks are FINISHED, but those FINISHED tasks should never be in the list of tasks to trigger. Thus here the tasks should not in FINISHED tasks logically. 

I'll update the original comment. And perhaps we could add an explicit `checkState(execution.getState() != FINISHED)` here ? If  we add this check we would also need to modify the test to exclude the FINISHED state.

Reply:
I guess you are referring to this part:
> Although we would also trigger checkpoint after tasks finished, but the finished tasks should not be in the list of tasks to trigger

As I understand it, the following checks is in `DefaultCheckpointPlanCalculator` **should not** be changed:
```
if (context.hasFinishedTasks() && !allowCheckpointsAfterTasksFinished) { throw exception }
checkAllTasksInitiated();
```

But his one **should**:
```
checkTasksStarted(result.getTasksToTrigger());
```

Reply:
Sorry for I just modified the comment and might cause some confusion... `Although we would also trigger checkpoint after tasks finished, but the finished tasks should not be in the list of tasks to trigger` This part is I just modified, and the original comment said `...to change the condition to only allows for RUNNING status or FINISHED status`.

I also think we should change the `checkTasksStarted(result.getTasksToTrigger());`. I think Till's concern should be why we not change it to `if (execution.getState() != ExecutionState.RUNNING && execution.getState() != ExecutionState.FINISHED )`

Reply:
Alright, so checking for `RUNNING` is enough, right @gaoyunhaii?

Reply:
Yes, checking for RUNNING is enough~